### PR TITLE
GODRIVER-1860 Only inspect errors messages with no codes

### PR DESCRIPTION
**Pull request by benjirewis**
[GODRIVER-1860](https://jira.mongodb.org/browse/GODRIVER-1860)

First commit syncs SDAM spec tests to add and update errors tests.

Remaining commits modify internally-used error helpers in `mongo` and `driver` to only inspect error messages when the error does not include a code (when the code is 0 by Go's int64 zero-value). Adds `10058` to list of not-master codes.

### CODE REVIEW

File: mongo/errors.go [hunk 1]
```diff
@@ -85,11 +85,10 @@ func replaceErrors(err error) error {
 
 // IsDuplicateKeyError returns true if err is a duplicate key error
 func IsDuplicateKeyError(err error) bool {
-	// handles SERVER-7164 and SERVER-11493
+	// handles SERVER-7164
 	for ; err != nil; err = unwrap(err) {
 		if e, ok := err.(ServerError); ok {
-			return e.HasErrorCode(11000) || e.HasErrorCode(11001) || e.HasErrorCode(12582) ||
-				e.HasErrorCodeWithMessage(16460, " E11000 ")
+			return e.HasErrorCode(11000) || e.HasErrorCode(11001) || e.HasErrorCode(12582)
```
Comment (benjirewis) [inline mongo/errors.go:91]: This internal use of `HasErrorCodeWithMessage` was a workaround for [SERVER-11493](https://jira.mongodb.org/browse/SERVER-11493). Duplicate key errors on a sharded cluster used to have an error code of 16460, a code common to a range of insert errors. So, we had to check the message as well. According to the discussion on 11493, though, sharded clusters now return a 11000 code for duplicate keys (backported to 2.4), so I think it's safe to remove this check. Note that I had to change the code in the associated test, too.

**Review by benjirewis — COMMENTED** [review]
Note that I’ve only updated error helpers that are _internally used_. I did not touch implementations of the helpers on the server interface: `HasMessage` and `HasCodeWithMessage`. My logic is that these helpers are only called in tests and are convenient tools for users. Ideally, a user would use the same behavior as the driver: not inspecting error messages without error codes. I could remove the message-based helpers, but as long as the logic within the Go driver does not rely on messages over codes, I think we’ve covered the spec. Let me know what you think.

**Comment by benjirewis** [comment]
It looks like `UnsupportedStorageEngine` error helpers ([here](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/errors.go#L86-L95) and [here](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/errors.go#L218-L221)) are an exception to the rule. The error code 20 represents an array of "Illegal Operations". We need to check for the "transaction numbers" prefix to know the error refers to an unsupported storage engine.

File: mongo/errors.go [hunk 2]
```diff
@@ -250,7 +249,8 @@ func (e CommandError) HasErrorCodeWithMessage(code int, message string) bool {
 
 // IsMaxTimeMSExpiredError returns true if the error is a MaxTimeMSExpired error.
 func (e CommandError) IsMaxTimeMSExpiredError() bool {
-	return e.Code == 50 || e.Name == "MaxTimeMSExpired"
+	noCode := e.Code == 0
```
Comment (LBWright) [inline mongo/errors.go:252]: Have you considered `hasCode` (or another typical bool identifier) as the variable name instead of `noCode`?

File: x/mongo/driver/errors.go [hunk 3]
```diff
@@ -317,12 +322,14 @@ func (e Error) NotMaster() bool {
...
-	return strings.Contains(e.Message, "not master")
+	noCode := e.Code == 0
+	return noCode && strings.Contains(e.Message, "not master")
 }
 
 // NamespaceNotFound returns true if this errors is a NamespaceNotFound error.
 func (e Error) NamespaceNotFound() bool {
-	return e.Code == 26 || e.Message == "ns not found"
+	noCode := e.Code == 0
+	return e.Code == 26 || noCode && e.Message == "ns not found"
```
Comment (iwysiu) [inline x/mongo/driver/errors.go:332]: add parentheses around `(noCode && e.Message == "ns not found")` to make the grouping more obvious

Comment (benjirewis) [reply x/mongo/driver/errors.go:332]: Sure! Added the same grouping for `IsMaxTimeMSExpiredError()`, as well.

File: mongo/errors.go (hunk 2 above)
Comment (benjirewis) [reply mongo/errors.go:252]: Thanks for your feedback @LBWright ! `hasCode` would require double negation (`hasCode := e.Code != 0; return !hasCode || ...`) which seems a little less readable. I've used `hasNoCode` instead to use a more bool-like name.

**Review by iwysiu — APPROVED** [review]

File: mongo/errors.go (hunk 1 above)
Comment (divjotarora) [reply mongo/errors.go:91]: I understand why we're making this change, but it seems out of scope for this ticket. The ticket is primarily about errors that convey SDAM state changes. Given that, I'd prefer to leave this code as-is, but let me know what you think.

File: x/mongo/driver/errors.go [hunk 4]
```diff
@@ -317,12 +322,14 @@ func (e Error) NotMaster() bool {
...
 	}
-	return strings.Contains(e.Message, "not master")
+	hasNoCode := e.Code == 0
+	return hasNoCode && strings.Contains(e.Message, "not master")
 }
 
 // NamespaceNotFound returns true if this errors is a NamespaceNotFound error.
 func (e Error) NamespaceNotFound() bool {
-	return e.Code == 26 || e.Message == "ns not found"
+	hasNoCode := e.Code == 0
```
Comment (divjotarora) [inline x/mongo/driver/errors.go:331]: I vaguely remember discussing whether or not we should change this error helper. What was the result of your investigation? Is this in scope for this ticket?

File: mongo/errors.go (hunk 1 above)
Comment (benjirewis) [reply mongo/errors.go:91]: It's a fair point. I would say that the general idea of the ticket is to prefer codes over messages in drivers generally, but it looks like other drivers are only making SDAM-related changes. So, I'll leave `IsDuplicateKeyError`, `IsMaxTimeMSExpiredError` and `NamespaceNotFound` as is.

File: x/mongo/driver/errors.go (hunk 4 above)
Comment (benjirewis) [reply x/mongo/driver/errors.go:331]: See comment above; changes to this helper removed.

**Review by divjotarora — APPROVED** [review]
